framework/db/schema/mssql/CMssqlCommandBuilder.php

```diff
@@ -224,6 +239,27 @@ protected function rewriteLimitOffsetSql($sql, $limit, $offset)
 		return $sql;
 	}
 
+	/**
+	 * Rewrite SQL to apply $limit and $offset for MSSQL database version 11 (2012) and newer.
+	 * @see https://learn.microsoft.com/en-us/sql/t-sql/queries/select-order-by-clause-transact-sql?view=sql-server-ver15#using-offset-and-fetch-to-limit-the-rows-returned
+	 * @see https://github.com/yiisoft/yii/issues/4491
+	 * @param string $sql sql query
+	 * @param integer $limit $limit > 0
+	 * @param integer $offset $offset > 0
+	 * @return string modified sql query applied w th limit and offset.
+	 */
+	protected function newRewriteLimitOffsetSql($sql, $limit, $offset)
+	{
+		$order = $this->findOrdering($sql);
+		if (empty($order)) {
+			$sql .= " ORDER BY (SELECT NULL)";
+		}
+
+		$sql .= sprintf(" OFFSET %d ROWS FETCH NEXT %d ROWS ONLY", $offset, $limit);
```

Comment:
This code is more compact and omits some conditions used in https://github.com/yiisoft/yii2/blob/364e907875fd57ee218085cca796ac5d1c3c8d51/framework/db/mssql/QueryBuilder.php#L88, is that on purpose?

Reply:
This function is being called from: https://github.com/yiisoft/yii/blob/8530c30959d5f68c0d1b0d611e3ce2bf32a8371a/framework/db/schema/mssql/CMssqlCommandBuilder.php#L195

If we look inside `applyLimit` we can see that `rewriteLimitOffsetSql` is only called when both `limit` and `offset` are set:
```php
if ($limit > 0 && $offset <= 0) //just limit
	$sql = preg_replace('/^([\s(])*SELECT( DISTINCT)?(?!\s*TOP\s*\()/i',"\\1SELECT\\2 TOP $limit", $sql);
elseif($limit > 0 && $offset > 0) // Limit and offset
	$sql = $this->rewriteLimitOffsetSql($sql, $limit,$offset);
```
As you can see the 'only offset' case is not handled, Causing `rewriteLimitOffsetSql` to only handle both limit and offset as opposed to https://github.com/yiisoft/yii2/blob/364e907875fd57ee218085cca796ac5d1c3c8d51/framework/db/mssql/QueryBuilder.php#L88, where it is possible for `limit` to be left out.

It turns out that querries only containing an `offset` were never generated for MSSQL. This could be fixed the same way as limit and offset have been fixed: by adding a `oldRewriteOffsetSql` and `newRewriteoffsetSql`, leaving `oldRewriteLimitOffsetSql` untouched.

 `oldRewriteOffsetSql($sql, $offset)` would generate the following SQL: 
```sql
SELECT *
FROM (
	SELECT *, 
	ROW_NUMBER() OVER (ORDER BY id) AS [__rowNumber__] -- Or order by (SELECT NULL) when no orderBy is given
	FROM [dbo].[item] [t]
) [t]
WHERE [__rowNumber__] > 15 -- Offset
```

And `newRewriteOffsetSql($sql, $offset)` would generate the following SQL
```sql
SELECT * FROM [dbo].[item]
ORDER BY id -- Or SELECT NULL) when no ordering is used
OFFSET 15 ROWS -- Offset
```